File: workers/PendingApprovalWorker.js
import pool from "../db.js";

export async function execute(context) {

  const { invoice_id, organization_id } = context;

  // Check for a payment approval decision for this invoice
  const decisionRes = await pool.query(
    `UPDATE invoice_payment_approvals
     SET processed = true
     WHERE id = (
       SELECT id
       FROM invoice_payment_approvals
       WHERE invoice_id      = $1
         AND organization_id = $2
         AND processed       = false
       ORDER BY decided_at DESC
       LIMIT 1
     )
     RETURNING id, decision, reason, reviewer_role, reviewer_name`,
    [invoice_id, organization_id]
  );

  if (decisionRes.rows.length) {
    return {
      success:       true,
      decisionFound: true,
      decision:      decisionRes.rows[0].decision,
      decisionId:    decisionRes.rows[0].id,
      reason:        decisionRes.rows[0].reason,
      reviewer_role: decisionRes.rows[0].reviewer_role,
      reviewer_name: decisionRes.rows[0].reviewer_name
    };
  }

  return {
    success:       true,
    decisionFound: false
  };
}

File: python/workers/pending_approval_worker.py
from db import get_connection

def run(state: dict) -> dict:
    invoice_id = state["invoice_id"]
    organization_id = state["organization_id"]

    # claim the latest unprocessed decision — atomic so a replayed event can't apply it twice
    conn = get_connection()
    cur = conn.cursor()
    cur.execute(
        """
        UPDATE invoice_payment_approvals
        SET processed = true
        WHERE id = (
            SELECT id FROM invoice_payment_approvals
            WHERE invoice_id = %s
            AND organization_id = %s
            AND processed = false
            ORDER BY decided_at DESC
            LIMIT 1
        )
        RETURNING decision, reviewer_name, reviewer_role
        """,
        (invoice_id, organization_id)
    )
    row = cur.fetchone()
    conn.commit()
    cur.close()
    conn.close()

    if not row:
        return {**state, "next_state": "PENDING_APPROVAL", "reason": "Waiting for payment approval decision"}

    decision, reviewer_name, reviewer_role = row

    if decision == "APPROVE":
        return {**state, "next_state": "ACCOUNTING", "reason": f"Payment approved by {reviewer_name} ({reviewer_role})"}

    if decision == "REJECT":
        return {**state, "next_state": "EXCEPTION_REVIEW", "reason": f"Payment rejected by {reviewer_name} ({reviewer_role}) — escalated for review"}

    return {**state, "next_state": "PENDING_APPROVAL", "reason": f"Unknown approval decision: {decision}"}
